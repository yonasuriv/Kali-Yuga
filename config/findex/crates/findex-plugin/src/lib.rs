#[cfg(feature = "findex_internals")]
pub mod findex_internal;

use abi_stable::std_types::*;

/// This struct is used to represent results by plugins and internal code of Findex.
#[derive(Clone)]
#[repr(C)]
pub struct FResult {
    /// Name of the result
    pub name: RString,
    /// Optional description of the result
    pub desc: ROption<RString>,
    /// The command to execute when the user presses enter
    pub cmd: ApplicationCommand,
    /// The icon of the result
    pub icon: RString,
    /// Score of the result. This will be used to sort multiple results
    pub score: isize,
}

#[derive(Clone, PartialEq)]
#[repr(C)]
pub enum ApplicationCommand {
    /// Exact command to execute
    Command(RString),
    /// AppId of GIO AppInfo
    Id(RString),
    /// Useful for results from plugins like calculator.
    None,
}

/// This macro is used to define a Findex plugin.
///
/// Example usage:
/// ```rust
/// use findex_plugin::{define_plugin, FResult};
/// use abi_stable::std_types::*;
///
/// fn init(config: &RHashMap<RString, RString>) -> RResult<(), RString>  {
///     // Set up your plugin using the config if necessary
///     // Return RErr if something went wrong
///
///     // Returning this indicates that the plugin initalization is successful
///     ROk(())
/// }
///
/// fn handle_query(query: RStr) -> RVec<FResult> {
///     let mut result = vec![];
///
///     /* Do stuff here */
///
///     RVec::from(result)
/// }
///
/// define_plugin!("prefix!", init, handle_query);
/// // or, add a shortcut key
/// define_plugin!("prefix!", "<Ctrl><Shift>p", init, handle_query);
/// ```
///
/// Refer to the `README.md` of this crate for more detailed explanation
#[macro_export]
macro_rules! define_plugin {
    ($prefix:literal, $init_function:ident, $query_handler:ident) => {
        #[no_mangle]
        #[used]
        pub static FINDEX_PLUGIN_PREFIX: &'static str = $prefix;

        #[no_mangle]
        extern "C" fn findex_plugin_init(
            config: &RHashMap<RString, RString>,
        ) -> RResult<(), RString> {
            $init_function(config)
        }

        #[no_mangle]
        extern "C" fn findex_plugin_query_handler(query: RStr) -> RVec<FResult> {
            $query_handler(query)
        }
    };

    ($prefix:literal, $keyboard_shortcut:literal, $init_function:ident, $query_handler:ident) => {
        #[no_mangle]
        #[used]
        pub static FINDEX_PLUGIN_PREFIX: &'static str = $prefix;

        #[no_mangle]
        #[used]
        pub static FINDEX_PLUGIN_KEYBOARD_SHORTCUT: &'static str = $keyboard_shortcut;

        #[no_mangle]
        extern "C" fn findex_plugin_init(
            config: &RHashMap<RString, RString>,
        ) -> RResult<(), RString> {
            $init_function(config)
        }

        #[no_mangle]
        extern "C" fn findex_plugin_query_handler(query: RStr) -> RVec<FResult> {
            $query_handler(query)
        }
    };
}
