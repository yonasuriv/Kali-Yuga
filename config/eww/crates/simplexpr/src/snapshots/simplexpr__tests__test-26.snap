---
source: src/lib.rs
expression: "p.parse(Lexer::new(\"foo.bar[2 + 2] * asdf[foo.bar]\"))"

---
Ok(
    BinOp(
        JsonAccess(
            JsonAccess(
                VarRef(
                    "foo",
                ),
                Literal(
                    "bar",
                ),
            ),
            BinOp(
                Literal(
                    "2",
                ),
                Plus,
                Literal(
                    "2",
                ),
            ),
        ),
        Times,
        JsonAccess(
            VarRef(
                "asdf",
            ),
            JsonAccess(
                VarRef(
                    "foo",
                ),
                Literal(
                    "bar",
                ),
            ),
        ),
    ),
)
